homeassistant/components/recorder/__init__.py

```diff
@@ -0,0 +1,358 @@
+        hass.bus.listen(MATCH_ALL, self.event_listener)
+
+    def run(self):
+        """Start processing events to save."""
+        from homeassistant.components.recorder.models import Events, States
+        import sqlalchemy.exc
+
+        while True:
+            try:
+                self._setup_connection()
+                self._setup_run()
+                break
+            except sqlalchemy.exc.SQLAlchemyError as e:
+                log_error(e, retry_wait=CONNECT_RETRY_WAIT, rollback=False,
+                          message="Error during connection setup: %s")
+
+        if self.purge_days is not None:
+            track_point_in_utc_time(self.hass,
+                                    lambda now: self._purge_old_data(),
+                                    dt_util.utcnow() + timedelta(minutes=5))
+
+        while True:
+            event = self.queue.get()
+
+            if event == self.quit_object:
+                self._close_run()
+                self.queue.task_done()
+                return
+
+            elif event.event_type == EVENT_TIME_CHANGED:
+                self.queue.task_done()
+                continue
+
+            for _ in range(0, RETRIES):
+                try:
+                    event_id = Events.record_event(Session, event)
+                    break
+                except sqlalchemy.exc.OperationalError as e:
+                    log_error(e, retry_wait=QUERY_RETRY_WAIT, rollback=True)
+
+            for _ in range(0, RETRIES):
+                try:
+                    if event.event_type == EVENT_STATE_CHANGED:
+                        States.record_state(
+                            Session,
+                            event.data['entity_id'],
+                            event.data.get('new_state'),
+                            event_id)
+                        break
+                except sqlalchemy.exc.OperationalError as e:
+                    log_error(e, retry_wait=QUERY_RETRY_WAIT, rollback=True)
+
+            self.queue.task_done()
+
+    def event_listener(self, event):
+        """Listen for new events and put them in the process queue."""
+        self.queue.put(event)
+
+    def shutdown(self, event):
+        """Tell the recorder to shut down."""
+        self.queue.put(self.quit_object)
+        self.block_till_done()
+
+    def block_till_done(self):
+        """Block till all events processed."""
+        self.queue.join()
+
+    def block_till_db_ready(self):
+        """Block until the database session is ready."""
+        self.db_ready.wait()
+
+    def _setup_connection(self):
+        """Ensure database is ready to fly."""
+        # pylint: disable=global-statement
+        global Session
+
+        import homeassistant.components.recorder.models as models
+        from sqlalchemy import create_engine
+        from sqlalchemy.orm import scoped_session
+        from sqlalchemy.orm import sessionmaker
+
+        _LOGGER.info("Recorder database URI %s", self.db_url)
```

Comment:
I don't think this is very useful. But if we want to keep it in we should reduce the level to `debug`.